ICE: layout: :evergreen_tree: : `Size != Size`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_transmute/src/layout/tree.rs:363:21: 'assertion `left == right` failed'', 'thread 'rustc' panicked at compiler/rustc_transmute/src/layout/tree.rs:363:21: 'assertion `left == right` failed''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
mod assert {
    use std::mem::{Assume, BikeshedIntrinsicFrom};

    pub fn is_maybe_transmutable<Src, Dst>()
    where
        Dst: BikeshedIntrinsicFrom<Src>,
    {
    }
}

fn distant_void() {
    enum Void {}

    enum DistantVoid {
        A(Void, u128),
    }

    assert::is_maybe_transmutable::<DistantVoid, ()>();
}


````

<details><summary><strong>original code</strong></summary>
<p>

original:
````rust
#![crate_type = "lib"]
#![feature(transmutability)]
#![allow(dead_code, incomplete_features, non_camel_case_types)]

mod assert {
    use std::mem::{Assume, BikeshedIntrinsicFrom};

    pub fn is_maybe_transmutable<Src, Dst>()
    where
        Dst: BikeshedIntrinsicFrom<Src, {
            Assume {
                alignment: true,
                lifetimes: true,
                safety: true,
                validity: true,
            }
        }>
    {}
}

fn void() {
    enum Void {}

    // This transmutation is vacuously acceptable; since one cannot construct a
    // `Void`, unsoundness cannot directly arise from transmuting a void into
    // anything else.
    assert::is_maybe_transmutable::<Void, u128>();

    assert::is_maybe_transmutable::<(), Void>(); //~ ERROR: cannot be safely transmuted
}

// Non-ZST uninhabited types are, nonetheless, uninhabited.
fn yawning_void() {
    enum Void {}

    struct YawningVoid(&'static Void);

    const _: () = {
        assert!(std::mem::size_of::<YawningVoid>() == std::mem::size_of::<u128>());
        // Just to be sure the above constant actually evaluated:
        assert!(false); //~ ERROR: evaluation of constant value failed
    };

    // This transmutation is vacuously acceptable; since one cannot construct a
    // `Void`, unsoundness cannot directly arise from transmuting a void into
    // anything else.
    assert::is_maybe_transmutable::<YawningVoid, u128>();

    assert::is_maybe_transmutable::<(), Void>(); //~ ERROR: cannot be safely transmuted
}

// References to uninhabited types are, logically, uninhabited, but for layout
// purposes are not ZSTs, and aren't treated as uninhabited when they appear in
// enum variants.
fn distant_void() {
    enum Void {}

    enum DistantVoid {
        A(Void, u128)
    }

    const _: () = {
        assert!(std::mem::size_of::<DistantVoid>() == std::mem::size_of::<usize>());
        // Just to be sure the above constant actually evaluated:
        assert!(false); //~ ERROR: evaluation of constant value failed
    };

    assert::is_maybe_transmutable::<DistantVoid, ()>();
    assert::is_maybe_transmutable::<(), Void>();
    assert::is_maybe_transmutable::<u128, DistantVoid>(); //~ ERROR: cannot be safely transmuted
}

````
</p>
</details>

Version information
````
rustc 1.81.0-nightly (0ef0dd245 2024-06-14)
binary: rustc
commit-hash: 0ef0dd24510b52da980889546fcd15254dc56a23
commit-date: 2024-06-14
host: x86_64-unknown-linux-gnu
release: 1.81.0-nightly
LLVM version: 18.1.7

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
warning: unused import: `Assume`
 --> /tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs:2:20
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                    ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs:19:2
   |
19 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs`

error[E0658]: use of unstable library feature 'transmutability'
 --> /tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs:2:20
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                    ^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information
  = help: add `#![feature(transmutability)]` to the crate attributes to enable
  = note: this compiler was built on 2024-06-14; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'transmutability'
 --> /tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs:2:28
  |
2 |     use std::mem::{Assume, BikeshedIntrinsicFrom};
  |                            ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information
  = help: add `#![feature(transmutability)]` to the crate attributes to enable
  = note: this compiler was built on 2024-06-14; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'transmutability'
 --> /tmp/icemaker_global_tempdir.YrqnXGtTwvMr/rustc_testrunner_tmpdir_reporting.h955EYnc6qOm/mvce.rs:6:14
  |
6 |         Dst: BikeshedIntrinsicFrom<Src>,
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #99571 <https://github.com/rust-lang/rust/issues/99571> for more information
  = help: add `#![feature(transmutability)]` to the crate attributes to enable
  = note: this compiler was built on 2024-06-14; consider upgrading it if it is out of date

thread 'rustc' panicked at compiler/rustc_transmute/src/layout/tree.rs:363:21:
assertion `left == right` failed
  left: Size(16 bytes)
 right: Size(0 bytes)
stack backtrace:
   0:     0x738fb540fae5 - std::backtrace_rs::backtrace::libunwind::trace::h66a8deddd31b55ec
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x738fb540fae5 - std::backtrace_rs::backtrace::trace_unsynchronized::h9b8c457e5846198e
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x738fb540fae5 - std::sys_common::backtrace::_print_fmt::h9692bfa044125fab
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x738fb540fae5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h8be930e4f4324fc1
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x738fb546077b - core::fmt::rt::Argument::fmt::h0147d2f1e77a5224
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/core/src/fmt/rt.rs:165:63
   5:     0x738fb546077b - core::fmt::write::h5bef47ac6c98803f
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/core/src/fmt/mod.rs:1168:21
   6:     0x738fb54046bf - std::io::Write::write_fmt::h7707c1b0ed1f3ef8
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/io/mod.rs:1835:15
   7:     0x738fb540f8be - std::sys_common::backtrace::_print::h79275c8f6c586b21
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x738fb540f8be - std::sys_common::backtrace::print::hd651502311c26a72
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x738fb54122f9 - std::panicking::default_hook::{{closure}}::hb66aeb12a772476e
  10:     0x738fb541209c - std::panicking::default_hook::hf6dbbcad84c9cd63
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/panicking.rs:292:9
  11:     0x738fb1cb38d0 - std[14120762d8ee0505]::panicking::update_hook::<alloc[303d001d1e4a23bb]::boxed::Box<rustc_driver_impl[2eb8f42bfae24b55]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x738fb5412bff - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h4b173c3d0381eea2
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/alloc/src/boxed.rs:2076:9
  13:     0x738fb5412bff - std::panicking::rust_panic_with_hook::h2897e2d711d501a3
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/panicking.rs:801:13
  14:     0x738fb5412827 - std::panicking::begin_panic_handler::{{closure}}::he291252945efa2fd
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/panicking.rs:667:13
  15:     0x738fb540ffa9 - std::sys_common::backtrace::__rust_end_short_backtrace::h9eb35470c0d5421a
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x738fb54124d4 - rust_begin_unwind
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/panicking.rs:658:5
  17:     0x738fb545cd33 - core::panicking::panic_fmt::h0c4e40ffed79f183
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/core/src/panicking.rs:74:14
  18:     0x738fb545d23e - core::panicking::assert_failed_inner::h4c5bb24f7c7d94d6
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/core/src/panicking.rs:410:17
  19:     0x738fb2872743 - core[f72d5f6884de128e]::panicking::assert_failed::<rustc_abi[83820ffeb3f18086]::Size, rustc_abi[83820ffeb3f18086]::Size>
  20:     0x738fb288727a - <rustc_transmute[e92ab8e1bc88ca7b]::layout::tree::Tree<rustc_transmute[e92ab8e1bc88ca7b]::layout::rustc::Def, rustc_transmute[e92ab8e1bc88ca7b]::layout::rustc::Ref>>::from_enum
  21:     0x738fb288a5e0 - <rustc_transmute[e92ab8e1bc88ca7b]::rustc::TransmuteTypeEnv>::is_transmutable
  22:     0x738fb38a83f1 - <rustc_trait_selection[140929a543f288fa]::traits::select::SelectionContext>::confirm_candidate
  23:     0x738fb38b336c - <rustc_trait_selection[140929a543f288fa]::traits::select::SelectionContext>::evaluate_candidate::{closure#0}::{closure#0}
  24:     0x738fb3432518 - <rustc_trait_selection[140929a543f288fa]::traits::select::SelectionContext>::evaluate_predicate_recursively::{closure#0}::{closure#0}
  25:     0x738fb342cc44 - rustc_traits[d1080c708c588edf]::evaluate_obligation::evaluate_obligation
  26:     0x738fb342c3e9 - rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::evaluate_obligation::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 2usize]>>
  27:     0x738fb342b952 - rustc_query_system[ef35600543bbfea8]::query::plumbing::try_execute_query::<rustc_query_impl[bf2ac6a6ab127c0b]::DynamicConfig<rustc_query_system[ef35600543bbfea8]::query::caches::DefaultCache<rustc_type_ir[d4d4d4b06d59bf13]::canonical::Canonical<rustc_middle[9ad05302be999b0c]::ty::context::TyCtxt, rustc_middle[9ad05302be999b0c]::ty::ParamEnvAnd<rustc_middle[9ad05302be999b0c]::ty::predicate::Predicate>>, rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 2usize]>>, false, false, false>, rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::QueryCtxt, false>
  28:     0x738fb342b5a4 - rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::evaluate_obligation::get_query_non_incr::__rust_end_short_backtrace
  29:     0x738fafdcb2f3 - <rustc_trait_selection[140929a543f288fa]::traits::fulfill::FulfillProcessor as rustc_data_structures[b6a8240b22e27d5b]::obligation_forest::ObligationProcessor>::process_obligation
  30:     0x738fb3008cc9 - <rustc_data_structures[b6a8240b22e27d5b]::obligation_forest::ObligationForest<rustc_trait_selection[140929a543f288fa]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[140929a543f288fa]::traits::fulfill::FulfillProcessor>
  31:     0x738faff8e3a9 - <rustc_hir_typeck[3a4328f626f4967a]::fn_ctxt::FnCtxt>::confirm_builtin_call
  32:     0x738fb3c48018 - <rustc_hir_typeck[3a4328f626f4967a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  33:     0x738fb3c42201 - <rustc_hir_typeck[3a4328f626f4967a]::fn_ctxt::FnCtxt>::check_block_with_expected
  34:     0x738fb3c48b57 - <rustc_hir_typeck[3a4328f626f4967a]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  35:     0x738fb31ae6ee - rustc_hir_typeck[3a4328f626f4967a]::check::check_fn
  36:     0x738fb31a4efe - rustc_hir_typeck[3a4328f626f4967a]::typeck
  37:     0x738fb31a497b - rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 8usize]>>
  38:     0x738fb3327f6e - rustc_query_system[ef35600543bbfea8]::query::plumbing::try_execute_query::<rustc_query_impl[bf2ac6a6ab127c0b]::DynamicConfig<rustc_query_system[ef35600543bbfea8]::query::caches::VecCache<rustc_span[b896b2121b1330a3]::def_id::LocalDefId, rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::QueryCtxt, false>
  39:     0x738fb3326a55 - rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  40:     0x738fb332668b - <rustc_middle[9ad05302be999b0c]::hir::map::Map>::par_body_owners::<rustc_hir_analysis[6ff8012e04425284]::check_crate::{closure#4}>::{closure#0}
  41:     0x738fb3325147 - rustc_hir_analysis[6ff8012e04425284]::check_crate
  42:     0x738fb3a68587 - rustc_interface[5734b4e3a95409b6]::passes::analysis
  43:     0x738fb3a680c7 - rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 1usize]>>
  44:     0x738fb3e498a5 - rustc_query_system[ef35600543bbfea8]::query::plumbing::try_execute_query::<rustc_query_impl[bf2ac6a6ab127c0b]::DynamicConfig<rustc_query_system[ef35600543bbfea8]::query::caches::SingleCache<rustc_middle[9ad05302be999b0c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[bf2ac6a6ab127c0b]::plumbing::QueryCtxt, false>
  45:     0x738fb3e4960f - rustc_query_impl[bf2ac6a6ab127c0b]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  46:     0x738fb3ce6bd2 - rustc_interface[5734b4e3a95409b6]::interface::run_compiler::<core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>, rustc_driver_impl[2eb8f42bfae24b55]::run_compiler::{closure#0}>::{closure#1}
  47:     0x738fb3d05089 - std[14120762d8ee0505]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[5734b4e3a95409b6]::util::run_in_thread_with_globals<rustc_interface[5734b4e3a95409b6]::util::run_in_thread_pool_with_globals<rustc_interface[5734b4e3a95409b6]::interface::run_compiler<core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>, rustc_driver_impl[2eb8f42bfae24b55]::run_compiler::{closure#0}>::{closure#1}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>::{closure#0}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>
  48:     0x738fb3d04e3a - <<std[14120762d8ee0505]::thread::Builder>::spawn_unchecked_<rustc_interface[5734b4e3a95409b6]::util::run_in_thread_with_globals<rustc_interface[5734b4e3a95409b6]::util::run_in_thread_pool_with_globals<rustc_interface[5734b4e3a95409b6]::interface::run_compiler<core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>, rustc_driver_impl[2eb8f42bfae24b55]::run_compiler::{closure#0}>::{closure#1}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>::{closure#0}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[f72d5f6884de128e]::result::Result<(), rustc_span[b896b2121b1330a3]::ErrorGuaranteed>>::{closure#2} as core[f72d5f6884de128e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x738fb541ca7b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb25f0033d53ed2e6
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/alloc/src/boxed.rs:2062:9
  50:     0x738fb541ca7b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hbb7141f8d9bd3912
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/alloc/src/boxed.rs:2062:9
  51:     0x738fb541ca7b - std::sys::pal::unix::thread::Thread::new::thread_start::hd182649926e2b713
                               at /rustc/0ef0dd24510b52da980889546fcd15254dc56a23/library/std/src/sys/pal/unix/thread.rs:108:17
  52:     0x738fae8a6ded - <unknown>
  53:     0x738fae92a0dc - <unknown>
  54:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.81.0-nightly (0ef0dd245 2024-06-14) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `(): core::mem::transmutability::BikeshedIntrinsicFrom<distant_void::DistantVoid, core::mem::transmutability::Assume { alignment: false, lifetimes: false, safety: false, validity: false }>`
#1 [typeck] type-checking `distant_void`
end of query stack
error: aborting due to 4 previous errors; 1 warning emitted

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [evaluate_obligation] evaluating trait selection obligation `(): core::mem::transmutability::BikeshedIntrinsicFrom<distant_void::DistantVoid, core::mem::transmutability::Assume { alignment: false, lifetimes: false, safety: false, validity: false }>`
#1 [typeck] type-checking `distant_void`

-->
@rustbot label +F-transmutability 


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"jswrenn"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_const_eval/src/interpret/discriminant.rs
+++ b/compiler/rustc_const_eval/src/interpret/discriminant.rs
@@ -241,7 +241,16 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         variant_index: VariantIdx,
     ) -> InterpResult<'tcx, Option<(ScalarInt, usize)>> {
         match self.layout_of(ty)?.variants {
-            abi::Variants::Single { .. } => Ok(None),
+            abi::Variants::Single { .. } => {
+                // The tag of a `Single` enum is like the tag of the niched
+                // variant: there's no tag as the discriminant is encoded
+                // entirely implicitly. If `write_discriminant` ever hits this
+                // case, we do a "validation read" to ensure the the right
+                // discriminant is encoded implicitly, so any attempt to write
+                // the wrong discriminant for a `Single` enum will reliably
+                // result in UB.
+                Ok(None)
+            }
 
             abi::Variants::Multiple {
                 tag_encoding: TagEncoding::Direct,

--- a/compiler/rustc_transmute/src/layout/tree.rs
+++ b/compiler/rustc_transmute/src/layout/tree.rs
@@ -341,37 +341,29 @@ pub(crate) mod rustc {
 
             // We consider three kinds of enums, each demanding a different
             // treatment of their layout computation:
-            // 1. enums that are uninhabited
-            // 2. enums for which all but one variant is uninhabited
-            // 3. enums with multiple inhabited variants
+            // 1. enums that are uninhabited ZSTs
+            // 2. enums that delegate their layout to a variant
+            // 3. enums with multiple variants
             match layout.variants() {
-                _ if layout.abi.is_uninhabited() => {
-                    // Uninhabited enums are usually (always?) zero-sized. In
-                    // the (unlikely?) event that an uninhabited enum is
-                    // non-zero-sized, this assert will trigger an ICE, and this
-                    // code should be modified such that a `layout.size` amount
-                    // of uninhabited bytes is returned instead.
-                    //
-                    // Uninhabited enums are currently implemented such that
-                    // their layout is described with `Variants::Single`, even
-                    // though they don't necessarily have a 'single' variant to
-                    // defer to. That said, we don't bother specifically
-                    // matching on `Variants::Single` in this arm because the
-                    // behavioral principles here remain true even if, for
-                    // whatever reason, the compiler describes an uninhabited
-                    // enum with `Variants::Multiple`.
-                    assert_eq!(layout.size, Size::ZERO);
+                Variants::Single { .. }
+                    if layout.abi.is_uninhabited() && layout.size == Size::ZERO =>
+                {
+                    // The layout representation of uninhabited, ZST enums is
+                    // defined to be like that of the `!` type, as opposed of a
+                    // typical enum. Consequently, they cannot be descended into
+                    // as if they typical enums. We therefore special-case this
+                    // scenario and simply return an uninhabited `Tree`.
                     Ok(Self::uninhabited())
                 }
                 Variants::Single { index } => {
-                    // `Variants::Single` on non-uninhabited enums denotes that
+                    // `Variants::Single` on enums with variants denotes that
                     // the enum delegates its layout to the variant at `index`.
                     layout_of_variant(*index)
                 }
                 Variants::Multiple { tag_field, .. } => {
                     // `Variants::Multiple` denotes an enum with multiple
-                    // inhabited variants. The layout of such an enum is the
-                    // disjunction of the layouts of its tagged variants.
+                    // variants. The layout of such an enum is the disjunction
+                    // of the layouts of its tagged variants.
 
                     // For enums (but not coroutines), the tag field is
                     // currently always the first field of the layout.

--- a/tests/ui/transmutability/enums/uninhabited_optimization.rs
+++ b/tests/ui/transmutability/enums/uninhabited_optimization.rs
@@ -19,8 +19,14 @@ enum SingleUninhabited {
     Y(Uninhabited),
 }
 
+enum MultipleUninhabited {
+    X(u8, Uninhabited),
+    Y(Uninhabited, u16),
+}
+
 fn main() {
     assert_transmutable::<Uninhabited>();
     assert_transmutable::<SingleInhabited>();
     assert_transmutable::<SingleUninhabited>();
+    assert_transmutable::<MultipleUninhabited>();
 }

--- a/tests/ui/transmutability/uninhabited.rs
+++ b/tests/ui/transmutability/uninhabited.rs
@@ -30,10 +30,32 @@ fn void() {
 }
 
 // Non-ZST uninhabited types are, nonetheless, uninhabited.
-fn yawning_void() {
+fn yawning_void_struct() {
     enum Void {}
 
     struct YawningVoid(Void, u128);
+
+    const _: () = {
+        assert!(std::mem::size_of::<YawningVoid>() == std::mem::size_of::<u128>());
+        // Just to be sure the above constant actually evaluated:
+        assert!(false); //~ ERROR: evaluation of constant value failed
+    };
+
+    // This transmutation is vacuously acceptable; since one cannot construct a
+    // `Void`, unsoundness cannot directly arise from transmuting a void into
+    // anything else.
+    assert::is_maybe_transmutable::<YawningVoid, u128>();
+
+    assert::is_maybe_transmutable::<(), Void>(); //~ ERROR: cannot be safely transmuted
+}
+
+// Non-ZST uninhabited types are, nonetheless, uninhabited.
+fn yawning_void_enum() {
+    enum Void {}
+
+    enum YawningVoid {
+        A(Void, u128),
+    }
 
     const _: () = {
         assert!(std::mem::size_of::<YawningVoid>() == std::mem::size_of::<u128>());

--- a/tests/ui/transmutability/uninhabited.stderr
+++ b/tests/ui/transmutability/uninhabited.stderr
@@ -7,10 +7,18 @@ LL |         assert!(false);
    = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error[E0080]: evaluation of constant value failed
-  --> $DIR/uninhabited.rs:65:9
+  --> $DIR/uninhabited.rs:63:9
    |
 LL |         assert!(false);
-   |         ^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: false', $DIR/uninhabited.rs:65:9
+   |         ^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: false', $DIR/uninhabited.rs:63:9
+   |
+   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0080]: evaluation of constant value failed
+  --> $DIR/uninhabited.rs:87:9
+   |
+LL |         assert!(false);
+   |         ^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: false', $DIR/uninhabited.rs:87:9
    |
    = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
 
@@ -36,11 +44,33 @@ LL | |             }
 LL | |         }>
    | |__________^ required by this bound in `is_maybe_transmutable`
 
-error[E0277]: `()` cannot be safely transmuted into `yawning_void::Void`
+error[E0277]: `()` cannot be safely transmuted into `yawning_void_struct::Void`
   --> $DIR/uninhabited.rs:49:41
    |
 LL |     assert::is_maybe_transmutable::<(), Void>();
-   |                                         ^^^^ `yawning_void::Void` is uninhabited
+   |                                         ^^^^ `yawning_void_struct::Void` is uninhabited
+   |
+note: required by a bound in `is_maybe_transmutable`
+  --> $DIR/uninhabited.rs:10:14
+   |
+LL |       pub fn is_maybe_transmutable<Src, Dst>()
+   |              --------------------- required by a bound in this function
+LL |       where
+LL |           Dst: BikeshedIntrinsicFrom<Src, {
+   |  ______________^
+LL | |             Assume {
+LL | |                 alignment: true,
+LL | |                 lifetimes: true,
+...  |
+LL | |             }
+LL | |         }>
+   | |__________^ required by this bound in `is_maybe_transmutable`
+
+error[E0277]: `()` cannot be safely transmuted into `yawning_void_enum::Void`
+  --> $DIR/uninhabited.rs:71:41
+   |
+LL |     assert::is_maybe_transmutable::<(), Void>();
+   |                                         ^^^^ `yawning_void_enum::Void` is uninhabited
    |
 note: required by a bound in `is_maybe_transmutable`
   --> $DIR/uninhabited.rs:10:14
@@ -59,7 +89,7 @@ LL | |         }>
    | |__________^ required by this bound in `is_maybe_transmutable`
 
 error[E0277]: `u128` cannot be safely transmuted into `DistantVoid`
-  --> $DIR/uninhabited.rs:70:43
+  --> $DIR/uninhabited.rs:92:43
    |
 LL |     assert::is_maybe_transmutable::<u128, DistantVoid>();
    |                                           ^^^^^^^^^^^ at least one value of `u128` isn't a bit-valid value of `DistantVoid`
@@ -80,7 +110,7 @@ LL | |             }
 LL | |         }>
    | |__________^ required by this bound in `is_maybe_transmutable`
 
-error: aborting due to 5 previous errors
+error: aborting due to 7 previous errors
 
 Some errors have detailed explanations: E0080, E0277.
 For more information about an error, try `rustc --explain E0080`.

